### ISSUE TITLE
Should we switch to Jekyll for our lessons?

### ISSUE THREAD

**Issue by gvwilson** [swcarpentry/DEPRECATED-lesson-template#279]
This issue continues the discussion started in http://lists.software-carpentry.org/pipermail/discuss_lists.software-carpentry.org/2015-June/003118.html - please add further comments on that thread here. In particular, please read [this post](http://lists.software-carpentry.org/pipermail/discuss_lists.software-carpentry.org/2015-June/003127.html) from @jdblischak on us having come full circle, and the cost of repeated changes to our build system, and [this one](http://lists.software-carpentry.org/pipermail/discuss_lists.software-carpentry.org/2015-June/003129.html) from @ocefpaf on using Travis-CI to build our HTML.

**Comment by cryptarch** [comment]
From the email discussion, I have a problems with @r-gaia-cs's defence of pandoc. Pandoc (and therefore R Studio) are not available in Arch Linux; even their Arch User Repos are broken. I've had serious problems contributing to R material and helping at R-based workshops because of this.

You might think, well, that's just one distro. However, it is an important and popular distro, and this problem with pandoc is furthermore symptomatic of a deeper problem: it is built on Haskell, which - while interesting and supported by many worthy nerds - is not robust in the slightest. We cannot rely on pandoc to not break in any other arbitrary distro or OS at any time.

I can only support a solution that is as robust and cross-platform as possible, with the least amount of dependencies or arbitrary "gotchas". That means dropping pandoc until such time as the Haskell community can get its act together on package management.

**Comment by ocefpaf** [comment]
Just to be clear, regarding my comment in the mailing list, if people need pandoc we can remove the burden of creating the HTML using `Travis-CI`.  Travis-CI does comes with a cost though.  (Complex configs, unexpected broken builds, convoluted token generation, etc.)  On the other hand, if people are OK leaving pandoc behind I do believe Jekyll is a better option.

PS: On an ~~un~~related note :stuck_out_tongue_winking_eye: , my favorite site generator is [Mkdocs](http://www.mkdocs.org/), but it is way to simple for SWC lessons :unamused:

**Comment by rgaiacs** [comment]
> Pandoc (and therefore R Studio) are not available in Arch Linux;

R Studio isn't available in others popular distro repositories. Users need to install it using the package provide at http://www.rstudio.com/products/rstudio/download/. Pandoc has similar issue and the package is provided at https://github.com/jgm/pandoc/releases.

I can try maintaining Pandoc for Arch after the summer.

> We cannot rely on pandoc to not break in any other arbitrary distro or OS at any time.

<del>I think this is true for any program.</del>

I understand your point and I think that we find a way to workaround it.

> I can only support a solution that is as robust and cross-platform as possible

<del>Are you saying that you will only support Java based solutions?</del>

Using Jekyll because of GitHub Pages **only** solve the problem because it is web based, i.e. you still have the problem to install Jekyll (or any other software) in some platforms.

**Comment by cryptarch** [comment]
> I can try maintaining Pandoc for Arch after the summer.

Haha, that would be awesome, although I don't think it will fix my underlying beef with Haskell, since I've also had xmonad break too many times ;)

> you still have the problem to install Jekyll (or any other software) in some platforms.

But this can be done per-user with gems. I've never had any problems with anything based on Ruby.

**Comment by cboettig** [comment]
At @gvwilson 's suggestion, posting my comments on kramdown here:

---

 tl;dr use redcarpet instead of kramdown.

If we tell people they can "use markdown" they probably think Github-flavored markdown, e.g. that they write fenced code blocks with language indicated:

``````
```python
import matplotlib
```
``````

which is identical to pandoc, and thus familiar to RStudio users and all existing lessons in pandoc.  I suspect users will be surprised to learn this doesn't work in Jekyll kramdown.  Most users Googling syntax highlighting in Jekyll will find things saying they should not put code in markdown at all, but use Jekyll's Liquid format [1], like so:

```
{% highlight python %}
import matplotlib
{% endhighlight %}
```

Further confusing things.  Least obvious might be the solution of "trailing styles"

```
~~~
import matplotlib
~~~
{: .language-ruby}
```

Is that really what we want?  I realize Jekyll is now an independent community project from GitHub with some historical legacy issues, but I have no idea why it continues to default to a syntax that is so divergent from Github's own markdown style.  Anyway, I would suggest we at least use redcarpet as the parser so that the code block notation we have been using in pandoc, Github, etc continues to work.  Trying to patch all knitr and these other tools to use trailing styles does not seem consistent with the goal of using familiar out-of-the-box tools.

**Comment by aaren** [comment]
One benefit of the new approach is that you can fork a lesson and make changes without leaving the browser, using Github's inbuilt editor.

I forked Greg's example repo and added some content, all using the browser: see https://github.com/gvwilson/using-jekyll/pull/3

This makes collaboration very easy.

**Comment by aaren** [comment]
@cboettig I don't like trailing styles either, but it appears that you can still specify the language of the code block as in GFM: http://kramdown.gettalong.org/syntax.html#language-of-code-blocks

**Comment by aaren** [comment]
@cboettig I looked at the code blocks a bit more. tldr: we can use the Kramdown GFM parser.

The default kramdown parser doesn't support backtick fences. It has a GFM parser (disabled but available by default) that does support them, and still appears to support trailing styles for the blockquotes (callouts etc.) - it does need hard wrapping disabling though. `_config.yml` needs to look like [this](https://github.com/aaren/using-jekyll/blob/211fd39b93bdc547c2a34798a257c7f5eed2136e/_config.yml).

We would still need the trailing styles for blockquote styling, which is unfortunate.

**Comment by gvwilson** [comment]
But blockquotes are much rarer than code blocks, so if we can style code 
blocks as most people are expecting, that will be a big help. I'll push 
a change to the using-jekyll repository to show this and see what people 
think.

**Comment by gvwilson** [comment]
Can we specify our own styles for code blocks on the opening line? In 
particular, can we specify 'out' and 'err' for stdout and stderr 
output?  Learners really like having these colorized so that they can 
tell what we type in from what the shell, SQLite, and other tools spit out.

**Comment by cboettig** [comment]
@gvwilson The notation

``````
```python
``````

in this setup will just make html like

```
<pre><code class='language-python'>
```

so you could just define CSS for that class to code it as input.  Output and error messages would have to be explicitly indicated as such in the markdown.

@gvwilson curious about your thoughts on syntax highlighting.  I think SWC has opted primarily for just coding all input text in one color and output text in another color.  I can only assume that research has shown that syntax highlighting helps spot errors; (though maybe that doesn't matter in this context?).  

It looks to me like syntax highlighting would have to be done by javascript in kramdown (which I think is why Jekyll documents the Liquid syntax instead; crazy mess).  Pandoc and redcarpet both generate markup around the actual code, allowing you to do syntax highlighting via css (e.g. using GitHub's css)

**Comment by gvwilson** [comment]
My concern is that if we turn on syntax highlighting and use a style 
that isn't recognized as a language (e.g., `err`), it'll bork.  I 
suppose I could just test it out...  Are you OK with the compromise of 
"standard style for code blocks, trailing style for blockquotes"?
Thanks,
Greg

**Comment by cboettig** [comment]
@gvwilson I'm okay with anything, I was just curious if the lack of syntax highlighting was based on theoretical objections or practical implementation ;-).

From my perspective, I imagine most people are familiar with how markdown looks and displays on Github, and so it seems ideal to emulate that as much as possible. redcarpet is essentially GFM, such that you can just drop in the github-light.css style sheet and things should just work.  

Github's Automatic Page Generator (https://help.github.com/articles/creating-pages-with-the-automatic-generator/) which does not use Jekyll is like this -- completely consistent with the markdown on Github.  Kramdown isn't, and introducing yet another flavor with it's esoteric markings that are different than the markdown we write anywhere else doesn't make sense to me.    

Redcarpet ships built in to jekyll, runs automatically on Github no travis nonsense or whatnot.  GFM _is_ essentially redcarpet: https://github.com/blog/832-rolling-out-the-redcarpet https://github.com/github/markup

Of course, regarding the input, output and error styling, GitHub' s sytnax highlighting implementation really isn't that great; and one might want to extend the css to make more explicit classes for output and errors.

**Comment by twhitehead** [comment]
> You might think, well, that's just one distro. However, it is an important and popular distro, and this problem with pandoc is furthermore symptomatic of a deeper problem: it is built on Haskell, which - while interesting and supported by many worthy nerds - is not robust in the slightest. We cannot rely on pandoc to not break in any other arbitrary distro or OS at any time.

@cryptarch I'm a bit surprised your experience with Haskell has been so bad. Is this a general thing with Haskell itself or primarily related to the issues you have had with the xmonad package? I ask as, although I've never used xmonad myself, I wouldn't be surprised it was a bit touchy being a low level unconventional piece of software (i.e., a niche tiling window manager with compile time configuration).

My experience over the past few years using pandoc under Debian for lecture notes (generating html, slides, and handouts from the same source) has been quite good (i.e., I can't recall it every having been broken). I would have expected the core of Haskell, that supported in the [Haskell Platform](https://www.haskell.org/platform/) on Mac, Windows, and Unix, to be quite stable.  Especially given the amount of [industry uptake](https://wiki.haskell.org/Haskell_in_industry).

**Comment by cryptarch** [comment]
@twhitehead It's been a few different things.

Part of it is just that Haskell package management in Arch Linux is a bit of a mess; as you can see [on the Arch wiki](https://wiki.archlinux.org/index.php/Haskell#Managing_Haskell_packages), there are multiple approaches which conflict each other. If you want the stability of the core repos, you suffer the price of not many packages, and even then I seem to recall the core xmonad package randomly breaking every now and then.

The non-core Arch Haskell repo would seem to be a good alternative except the last time I checked (a few months ago) it no longer offers Pandoc.

Or perhaps one could just install Pandoc as a user into your home directory and not worry about having it as a system package, but then since cabal is not a package manager this always seems to result in conflicts.

Aside from the package management situation, I find certain aspects of the language itself a bit problematic, although probably not a deal breaker by themselves. For example, compare with R. In R, if a function uses too much memory, R will kill the function and tell you you're out of memory, but other than that everything's safe. In Haskell, an elementary task is building a vector that can contain a potentially  infinite number of elements. If you then ask for the length of that vector, Haskell will gobble up all your memory and crash your computer without warning.

If I have to be worrying about computer memory, then I'd rather work in a language that lets me talk to the computer directly, such as C; C doesn't pretend that hardware constraints are not a problem. Haskell on the other hand seems to think that physical computational machines are beneath its notice.
